### PR TITLE
feat: add support to KFDDistribution kind

### LINKED ISSUE #384: Support `KFDDistribution` kind configuration file

### PR DESCRIPTION
**Pull request by alessiodionisi**
Closes #384.

SSIA

⚠️ Merge after https://github.com/sighupio/fury-distribution/pull/95 and update `go.mod` dependency

### CODE REVIEW

**Review by nutellinoit — CHANGES_REQUESTED** [review]
While testing we found two things:

1) We should check that at least a storageClass exists in the cluster with `kubectl get sc`
2) We should also thing as a general change to do the first apply with `kustomize build . | yq 'select(.kind == "CustomResourceDefinition")'`

**Comment by nutellinoit** [comment]
When running this command: `./furyctl create cluster --distro-location /Users/samuelechiocca/github.com/sighupio/fury-distribution`

Screenshot 2023-06-06 at 10 14 53

to overcome this i needed `--skip-deps-validation`

**Comment by nutellinoit** [comment]
We should also check if all the nodes in the cluster where we are installing the distro are `Ready`, otherwise it means that there is no CNI

**Comment by alessiodionisi** [comment]
Pushed commits that fixes the validation error, checks storage class and node readiness.

**Comment by smerlos** [comment]
---
Deploying KFDDistribution Cluster: Pre-conditions and Recommendations
---

## Pre-conditions for Deploying Cluster

Before deploying your cluster using `kind: KFDDistribution`, make sure to fulfill the following pre-conditions:

1. **All cluster nodes must be ready**: Ensure that each node in your Kubernetes cluster is in the 'Ready' state. Refer to the [Kubernetes node documentation](https://kubernetes.io/docs/concepts/architecture/nodes/) for additional guidance.
   
2. **A CNI (Container Network Interface) must be installed**: The cluster should be set up with a CNI that suits your use case. 

3. **A Storage Provider must be configured**: This involves setting up a `StorageClass` for your cluster. For more information, check out the [StorageClass documentation](https://kubernetes.io/docs/concepts/storage/storage-classes/).

## Recommendations

Here are some recommendations to optimize your KFDDistribution configuration:

- Consider configuring or removing `spec.distribution.common.[nodeSelector, tolerations]` based on your use case.

- Depending on your Kubernetes provider, be mindful of Gatekeeper policies. Add any namespaces necessary for the functioning of basic components (like the Storage Provider, CNI) or any other necessary namespaces, apart from `kube-system`, to the list of excluded namespaces (`spec.distribution.modules.policy.gatekeeper.additionalExcludedNamespaces`). Example: with Kind, you should add `local-path-storage` to the list of excluded namespaces.

Please feel free to modify and adapt these instructions based on your unique use case and technical requirements.

**Comment by alessiodionisi** [comment]
⚠️ Merge after https://github.com/sighupio/fury-distribution/pull/95 and update `go.mod` dependency

**Review by Al-Pragliola — CHANGES_REQUESTED** [review]

**Comment by alessiodionisi** [comment]
rebased

File: internal/template/config.go [hunk 1]
```diff
@@ -67,6 +67,36 @@ func NewConfig(tplSource, data *merge.Merger, excluded []string) (Config, error)
 	return cfg, nil
 }
 
+func NewConfigWithoutData(tplSource *merge.Merger, excluded []string) (Config, error) {
+	var cfg Config
+
+	if *tplSource.GetCustom() == nil {
```
Comment (omissis) [inline internal/template/config.go:73]: this is a potential null pointer exception, chack for tplSource to be not nil
```suggestion
	if tplSource == nil || *tplSource.GetCustom() == nil {
```

**Review by omissis — APPROVED** [review]

**Pull request merged**